### PR TITLE
Support cast between Durations + between Durations all numeric types

### PR DESCRIPTION
**Pull request by tisonkun**
# Which issue does this PR close?

This resolves https://github.com/apache/arrow-rs/discussions/6425.

# Rationale for this change
 
Durations are represent by i64. They can be casted from/to each other just like Timestamps.

# What changes are included in this PR?

Support cast between Durations. And catch up support cast between Durations and all numeric type.

# Are there any user-facing changes?

Not quite. Users would found new reasonable cast supported, nothing works currently changed.

### CODE REVIEW

**Comment by tisonkun** [comment]
cc @alamb @tustvold Here is the patch.

I'm finding where I can add some tests. If you can point me out it would be quite helpful ..

**Comment by tisonkun** [comment]
CI failure seems exist to master also. Likely unrelated to this patch ..

**Comment by alamb** [comment]
> CI failure seems exist to master also. Likely unrelated to this patch ..

I believe it is https://github.com/apache/arrow-rs/issues/6448 which was subsequently fixed

File: arrow-cast/src/cast/mod.rs [hunk 1]
```diff
@@ -5254,6 +5286,23 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_cast_between_durations() {
```
Comment (alamb) [inline arrow-cast/src/cast/mod.rs:5290]: Can you please add additional tests for:
1. Casting between each individual duration type (cover second, millisecond, microsecond, and nanosecond)
2. Tests where the casting will fail (aka when checked_mul fails)
3. Casting to/from numeric types to Duration types (to ensure that we don't break the feature by accident in the future)

**Review by alamb — COMMENTED** [review]
Thanks @tisonkun  -- this looks very nice 🙏 

I think this PR just needs a few more tests and it will be good to merge

File: arrow-cast/src/cast/mod.rs (hunk 1 above)
Comment (tisonkun) [reply arrow-cast/src/cast/mod.rs]: Updated.

**Review by alamb — APPROVED** [review]
Looks great to me -- thank you @tisonkun

**Review by waynexia — APPROVED** [review]
Looks good to me, thanks @tisonkun

**Pull request merged**